v2/spiffetls/spiffetls_test.go

```diff
@@ -0,0 +1,472 @@
+			dialMode:   spiffetls.MTLSClientWithSourceOptions(tlsconfig.AuthorizeID(serverID), workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.MTLSServerWithSourceOptions(tlsconfig.AuthorizeID(clientID), workloadapi.WithClientOptions(workloadapi.WithAddr(wlAPIServerA.Addr()))),
+		},
+		{
+			name:                "MTLSWebClientWithSourceOptions / MTLSWebServerWithSourceOptions succeeds",
+			dialMode:            spiffetls.MTLSWebClientWithSourceOptions(webCertPool, workloadapi.WithClient(wlAPIClientB)),
+			listenMode:          spiffetls.MTLSWebServerWithSourceOptions(tlsconfig.AuthorizeID(clientID), webCert, workloadapi.WithClientOptions(workloadapi.WithAddr(wlAPIServerA.Addr()))),
+			clientConnPeerIDErr: "spiffetls: no URI SANs",
+		},
+
+		// *WithRawConfig Scenario
+		{
+			name:                "TLSClientWithRawConfig / TLSServerWithRawConfig succeeds",
+			dialMode:            spiffetls.TLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), bundleSource),
+			listenMode:          spiffetls.TLSServerWithRawConfig(wlAPISourceA),
+			serverConnPeerIDErr: "spiffetls: no peer certificates",
+		},
+		{
+			name:       "MTLSClientWithRawConfig / MTLSServerWithRawConfig succeeds",
+			dialMode:   spiffetls.MTLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), svidSource, bundleSource),
+			listenMode: spiffetls.MTLSServerWithRawConfig(tlsconfig.AuthorizeID(clientID), wlAPISourceA, bundleSource),
+		},
+		{
+			name:                "MTLSWebClientWithRawConfig / MTLSWebServerWithRawConfig succeeds",
+			dialMode:            spiffetls.MTLSWebClientWithRawConfig(webCertPool, svidSource),
+			listenMode:          spiffetls.MTLSWebServerWithRawConfig(tlsconfig.AuthorizeID(clientID), webCert, bundleSource),
+			clientConnPeerIDErr: "spiffetls: no URI SANs",
+		},
+	}
+
+	testClose(t, spiffetls.TLSServerWithSource(wlAPISourceA), spiffetls.TLSClientWithSource(tlsconfig.AuthorizeID(serverID), wlAPISourceB))
+
+	for _, test := range tests {
+		test := test
+
+		if test.defaultWlAPIAddr != "" {
+			require.NoError(t, os.Setenv("SPIFFE_ENDPOINT_SOCKET", test.defaultWlAPIAddr))
+		} else {
+			require.NoError(t, os.Unsetenv("SPIFFE_ENDPOINT_SOCKET"))
+		}
+
+		t.Run(test.name, func(t *testing.T) {
+			// Start listening
+			listenCtx, cancelListenCtx := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancelListenCtx()
+
+			var wg sync.WaitGroup
+			var listener net.Listener
+			var listenAddr string
+			listenDataCh := make(chan string, 1)
+			listenErrCh := make(chan error, 1)
+			if test.listenLAddr == "" {
+				test.listenLAddr = defaultListenLAddr
+			}
+			if test.listenProtocol == "" {
+				test.listenProtocol = defaultListenProtocol
+			}
+			if test.listenMode != nil {
+				listener, err = spiffetls.ListenWithMode(listenCtx, test.listenProtocol, test.listenLAddr, test.listenMode, test.listenOption...)
+				if test.listenErr != "" {
+					require.Error(t, err)
```

Comment:
nit: `EqualError` will cover the case where the err is nil so this isn't strictly necessary